File: Kern/Sources/DuoKern/Box/Exposure.swift
import Foundation

extension BoxEngine {

    /// Cards worth surfacing for passive exposure (home/lock-screen widgets,
    /// watch complications), most-urgent first and fully deterministic. Tiers:
    ///
    ///   0. relearning — just lapsed, reinforce first
    ///   1. new & queued — prime a word before its first study
    ///   2. learning — mid-acquisition
    ///   3. review — weakest (lowest stability) first
    ///   4. upcoming — next automatic new words in seed order, so the list is
    ///      never empty and previews what's coming
    ///
    /// Only the review tier has a memory to rank by; new/queued/upcoming cards
    /// have none, so they keep queue- resp. seed-order. Suspended cards and
    /// locked phrases are excluded. Pure: `now` is unused today but kept for
    /// parity with the rest of the engine and future due-weighting.
    public static func exposureCards(state: BoxState, now: Date, limit: Int) -> [Card] {
        let direction = state.config.direction
        var ranked: [(tier: Int, order: Double, id: String, card: Card)] = []

        // Introduced, active cards → tier by phase, order by stability.
        for sched in state.scheduling.values
            where sched.direction == direction && !sched.suspended && sched.memory != nil {
            guard let card = state.cards[sched.cardID] else { continue }
            let tier: Int
            switch sched.phase {
            case .relearning: tier = 0
            case .learning:   tier = 2
            default:          tier = 3   // review
            }
            ranked.append((tier, sched.memory?.stability ?? 0, sched.cardID, card))
        }

        // Queued-but-unstudied new cards → tier 1, in queue order.
        for (index, id) in enqueuedEligible(state).enumerated() {
            if let card = state.cards[id] { ranked.append((1, Double(index), id, card)) }
        }

        // Preview: the next automatic new words in seed order → tier 4.
        let alreadyRanked = Set(ranked.map(\.id))
        let upcoming = seedOrdered(state.cards.values.filter { card in
            scheduling(state, card.id) == nil && !alreadyRanked.contains(card.id)
                && (card.kind != .phrase || card.componentIDs.isEmpty || isPhraseUnlocked(state, card))
        }).prefix(limit)
        for (index, card) in upcoming.enumerated() {
            ranked.append((4, Double(index), card.id, card))
        }

        return ranked
            .sorted { ($0.tier, $0.order, $0.id) < ($1.tier, $1.order, $1.id) }
            .prefix(limit)
            .map(\.card)
    }
}

File: kern/src/commonMain/kotlin/net/spross/kern/box/Exposure.kt
package net.spross.kern.box

import net.spross.kern.model.Card
import net.spross.kern.model.CardPhase

internal object Exposure {

    private data class Ranked(val tier: Int, val order: Double, val card: Card)

    /**
     * Cards worth surfacing for passive exposure (widgets, watch), most urgent first.
     * v1's tiers:
     *   0 relearning · 1 enqueued-eligible new (queue order) · 2 learning ·
     *   3 review (weakest stability first) · 4 upcoming introducible cards in seed order.
     * Display surfaces always render the TARGET realization.
     *
     * [eligible] rejects cards a surface cannot render — it gates before the
     * limit, so a surface still gets a full [limit] of the cards it can show.
     */
    fun exposureCards(state: BoxState, limit: Int, eligible: (Card) -> Boolean = { true }): List<Card> {
        if (limit <= 0) return emptyList()
        val ranked = mutableListOf<Ranked>()

        for (sched in Inventory.scheduled(state)) {
            if (sched.suspended || sched.memory == null) continue
            if (!eligible(state.cards.getValue(sched.cardId))) continue
            val tier = when (sched.phase) {
                CardPhase.Relearning -> 0
                CardPhase.Learning -> 2
                else -> 3
            }
            ranked += Ranked(tier, sched.memory.stability, state.cards.getValue(sched.cardId))
        }

        Growth.enqueuedEligible(state)
            .map { state.cards.getValue(it) }
            .filter(eligible)
            .forEachIndexed { index, card ->
                ranked += Ranked(1, index.toDouble(), card)
            }

        // Preview: the next introducible cards in seed order, so the list is never empty.
        val rankedIds = ranked.mapTo(mutableSetOf()) { it.card.id }
        Inventory.joinedCards(state)
            .filter {
                state.scheduling[it.id] == null && it.id !in rankedIds &&
                    Growth.isIntroducible(state, it) && eligible(it)
            }
            .take(limit)
            .forEachIndexed { index, card ->
                ranked += Ranked(4, index.toDouble(), card)
            }

        return ranked
            .sortedWith(compareBy({ it.tier }, { it.order }, { it.card.id }))
            .take(limit)
            .map { it.card }
    }
}
